Inherit Implementation of type outside current crate - Internal Compile Error

## Inherit Implementation of type outside current crate:

- You can view the same on godbolt [`here`](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATKUcAZPAZMADl3ACNMYhBpAAdUBUI7Bhc3D29SOITbAQCg0JYIqOkrTBskoQImYgIU908fErKBCqqCXJDwyOjLSuratIbetsCOgq7JAEpLVBNiZHYOAFIvAGYwMEWAVgAhMVpUAHcIEwYzTHQJrYARRY0AQSoGAGoWJkCICafFgHZt27u8CwYrQngA1UqLFZOEwADkh2C%2Bv0RVyeAHpUU9IsQSP8flcOFNaJxNrxPBwtKRUJwAEpmAhPBQzOaYL6rHikAiaAlTADWIBWGgAdBouCsuBoAGxcTYAThhMsk330nEkpK5lM4vAUIA0HK5UzgsBgiBQqCBdEi5EoaHN9CiwGQyGICgAtApmDEFAhUBYsAA3PDzUF4TAHADyMUYnHZNFoBEi2ogYXVYUCVQAntHeKnmMR02GwtpSpzuLwbWxBGGGLRM%2BTeFgwiZgE49trS6QsK8jOI6x28MRi3g/Zg2xTMKpSiZ41nyIJMETe7Q8GFiBmXFh1QRiICs1MqAZgApg6GI1H2/xBCIxOwpDJBIoVOpe7ouPpDMY6fpl9rIFNUDFsgYNteFQYdiG3LAfygZg2BAc5EgEUg/TEEwFi8DQvB4CYpkaQCHAYZxXDqPR/BGfJCj0TIEOSIi0lfKjAPaciulfXDyiGfpPFY%2BdB2aIYmM6KJWI42iuJ6VoBLGIScKZeYJEJYk1V7KkOCeVQYQlF0JUkJ4AHEnCcakhCeCB9MM4zXnMSJPggXBCBIVkxQmXgSy0bDSD5TZdQXVVSBYaIZUFGFNglTCuAlFYJQ0YKZQlUgyQpFStR1PU6wNY0ICQGYCBiKcrQgG0YgtYhglYBZTHMZAni4QUJRhYVeHgogIL0C9hFEcRbzah81HVF9SAOVcYl3ZUOBJeL1RUsMp1y%2BlUCoNSNK0nSzKMkzVospgrOIGyXFtSJHK4ZzUrchSOF8/yvMFKQVm%2BGEYUilYZSkDRpASkDNUsFLXO5DzvHqzZvikLxgqByQQakJUFxWJTEs%2Bn70vgQ0TUK4r8tRu0QGIKQZV1f1A0wY9w0jMkYzoeNnUoZNexzDMZ1pvMCyLGwZ3LRgCCrGt1QbJsW1oWg23ZTt3x7Cl8AHMph1HRqJ2QKcFnZQJ4wXCklxXNcMAWCktx3Us9wPI8Q2Js8Y1kK9OukbqlF6590gMIwsc/NWoL/ACkmAykwIgkd4Bwnimk8CBHE4kiCMkij6PiaiQ8jrIknDljLH9wCWj6US9DYviJLIwSM5E1IxPMbO8lzo7plmOSy4Xcb3o1VSKoIKqaskQKNBMuzmsO46EamBBMCYLAog%2BP7xW8lVeEu3Va6Sr7dQR0hkcykBstm9GzSKu1Stgjh1M07S9IMtbTMPzbtuOpqSDwdBWrNjqb0t2QeqfCldB8QamGGvXRprybOGmnKpxPHmk8BuTdBQt2FCZPaG8DrLBWF4bu%2Bpe790HpQXkEgNBj3OhPEAXkJrKU%2BtqOeSC/qSGbtKSQEpNheDqhKb4QMVijRhvguGHAXIkIXF4WGH02EnV%2BmBBI9hJBAA%3D) & [`here`](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgATKUcAZPAZMADl3ACNMYhAANlIAB1QFQjsGFzcPb3jE5IEAoNCWCKjYq0wbFKECJmICNPdPH1LygUrqgjyQ8MiYyyqauozGvvbAzsLu6IBKS1QTYmR2DgBSLwBmMDAlgFYAITFaVAB3CBMGM0x0Se2AESWNAEEqBgBqFiZAiEnnpYB2Hbv7gQAJ5xTDPHZAgiYBTfVbXZ4ANTKS1WThMAA4UdgUf8HgC8Cw4rRwZDod8/uT4QB6KnPSLEEikZ4KVhgpgwphhVAAN0wAN%2B1w401onC2vE8HC0pFQnAASmYCMzZvMwStVjxSARNELpgBrECrDQAOg0XHVGmiXC2AE50dbJD99JxJOLtdLOLwFCANJrtdM4LAYIgUKhCXRIuRKGgw/QosBkMhiAoALQKZhxBQIVAWLDcvALBF4TCHADyoIlGpotChScoYTdYUC1SBnA1jeYxCBJbC2jKWu4vGjbEEJYYtBbkt4WDCJmATn2XoHpCwbyM4kny7wxD7eF5i6lmFUZRMUNbvECUJFG9oeDCxGbLiwboIxAJZ%2BmVAMwAUheLZcYZ4yIIIhiOwUhAfIShqG6uhcPohjGAq%2Bi3l6kDTKgcS2AIi68Dy9J4FgqFQMwbAgBcOQMKQ3JiCYixeBoXg8JM0xNFhngQI4AyeHB/ijAURR6AkSRsVxgnZGxHT8d0cGsRUwyiTJmDWGxrQ1JJXRRDJ8muPUejmG06njJpLHKgsEjCqKrobjKHDPKo6LRMm0SSM8ADiThOLKQjPBA7med5bzmJEXwQLghAkN8axcJMvD9lozGkPqWw%2BleLqkCwICSNaRrols0SMVw0SrNEGi5dasQSlKNmet6vqTv6QYQEgswEHEJ6RhA0ZxOGxDBKynCmOYyDPFwRrROiJq8ORRCvugej8MBojiOBC2QSo6gbrBpCHPecTvk6HBiqQlW4ZwJYnm1iqoFQdkOU5Ll%2BV5PmPQF7I1iFLgxpEkXqjFdXxRZHBpRlyVGlIqw/Oi6LFas1pSBo0gne6HA1T6cU6ol3gTVsPxSF4uW45I%2BNSI6V6rFZVUev9QqkAGUCNUgXU9R1TOxiAxBSNaPq5vmmC/qW5aAVWNZehA9Ybu2zaAZLnbdr2NiAUOjAEKO45utOs7zrQtCLhqK4IeuUr4Nu5R7m6h7HqeS4Xkpbo3neD4YIsUovm%2BA4fl%2BP5FgLAFLqtIHLdIq2KOtMGZAYRjs0h9tEehmEpDh0q8sQs3QvALFKTu9gcQwzg6RkPG54ZAlwUJFEKVkwkpMX0mWJnzQMKptT59xdfKXJBl8Rpenaekrf6WpXdGeZMxzGZ0UHUdSM2c8g0EMNo1ZSaPlhTNP3RbFfrTAgmBMFgUSfJjXAaClzq8CDPrT1TXpo36tMMyALWXSzobdbGfWkRw9mOc5bkeU9vl/6vSCsQP600SAEXmrIAOYEg6yBDtBTamQdpMD2u7SeFNTocHOq1E8zxrqzwVAvI0S8NA%2BU%2Bm/b6aovB/XRglHee9uiH31MfU%2BQNz4gGSsdN01VLC1ToXqTKi8rSSGiFsLw41og/FxqsA65NuHWSpgIg6XhMHI03vVaYyckj2EkEAA%3D%3D%3D)

---

### I tried this code from [`E0116`](https://doc.rust-lang.org/error_codes/E0116.html):

```rust
#![allow(unused)]
fn main() {
impl Vec<u8> { } // error
}
```
- Also, this
```rust
#![allow(unused)]
fn main() {
type Bytes = Vec<u8>;

impl Bytes { } // error, same as above
}
```

---

### I Expected to see this happen:
- GIve error like rustc.
 - i.e., 
   - `error[E0116]: cannot define inherent impl for a type outside of the crate where the type is defined` 

---

### Instead this happened:
- First, it gives `failed to resolve TypePath: x in this scope`
- Then, internal compiler error. 

```bash
➜  gccrs-build gcc/crab1 ../mahad-testsuite/E0116A.rs 
../mahad-testsuite/E0116A.rs:3:6: error: failed to resolve TypePath: Vec<Type: u8> in this scope
    3 | impl Vec<u8> { } // error
      |      ^~~
crab1: internal compiler error: in visit, at rust/resolve/rust-ast-resolve-item.cc:574
0x7b8ce4 Rust::Resolver::ResolveItem::visit(Rust::AST::InherentImpl&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-item.cc:574
0xbb2167 Rust::Resolver::ResolveItem::go(Rust::AST::Item*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-item.cc:212
0xbc2bf8 Rust::Resolver::ResolveStmt::go(Rust::AST::Stmt*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-stmt.h:43
0xbc2bf8 Rust::Resolver::ResolveStmt::go(Rust::AST::Stmt*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-stmt.h:35
0xbc2bf8 Rust::Resolver::ResolveExpr::visit(Rust::AST::BlockExpr&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-expr.cc:316
0xbc4aaf Rust::Resolver::ResolveExpr::go(Rust::AST::Expr*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&, bool)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-expr.cc:36
0xbb64f6 Rust::Resolver::ResolveItem::visit(Rust::AST::Function&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-item.cc:540
0xbb2167 Rust::Resolver::ResolveItem::go(Rust::AST::Item*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-item.cc:212
0xba68fa Rust::Resolver::NameResolution::go(Rust::AST::Crate&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve.cc:108
0xa6a839 Rust::Session::compile_crate(char const*)
	../../gccrs/gcc/rust/rust-session-manager.cc:595
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
```
- For Second Example:
```bash
➜  gccrs-build gcc/crab1 ../mahad-testsuite/E0116B.rs
../mahad-testsuite/E0116B.rs:5:6: error: failed to resolve TypePath: Bytes in this scope
    5 | impl Bytes { } // error, same as above
      |      ^~~~~
crab1: internal compiler error: in visit, at rust/resolve/rust-ast-resolve-item.cc:574
0x7b8ce4 Rust::Resolver::ResolveItem::visit(Rust::AST::InherentImpl&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-item.cc:574
0xbb2167 Rust::Resolver::ResolveItem::go(Rust::AST::Item*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-item.cc:212
0xbc2bf8 Rust::Resolver::ResolveStmt::go(Rust::AST::Stmt*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-stmt.h:43
0xbc2bf8 Rust::Resolver::ResolveStmt::go(Rust::AST::Stmt*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-stmt.h:35
0xbc2bf8 Rust::Resolver::ResolveExpr::visit(Rust::AST::BlockExpr&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-expr.cc:316
0xbc4aaf Rust::Resolver::ResolveExpr::go(Rust::AST::Expr*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&, bool)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-expr.cc:36
0xbb64f6 Rust::Resolver::ResolveItem::visit(Rust::AST::Function&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-item.cc:540
0xbb2167 Rust::Resolver::ResolveItem::go(Rust::AST::Item*, Rust::Resolver::CanonicalPath const&, Rust::Resolver::CanonicalPath const&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve-item.cc:212
0xba68fa Rust::Resolver::NameResolution::go(Rust::AST::Crate&)
	../../gccrs/gcc/rust/resolve/rust-ast-resolve.cc:108
0xa6a839 Rust::Session::compile_crate(char const*)
	../../gccrs/gcc/rust/rust-session-manager.cc:595
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
```


---

### Meta

- What version of Rust GCC were you using, git sha b75357f4a1835dcd288df67402340b4cb6b96c12.
- Godbolt version:
   -  gccrs (Compiler-Explorer-Build-gcc-a06cec181de6398231e0662c1ed83d905ad6d1d9-binutils-2.40) 13.0.1 20230417 (experimental)

---

--- a/gcc/rust/resolve/rust-ast-resolve-item.cc
+++ b/gcc/rust/resolve/rust-ast-resolve-item.cc
@@ -582,7 +582,14 @@ ResolveItem::visit (AST::InherentImpl &impl_block)
   // Setup paths
   CanonicalPath self_cpath = CanonicalPath::create_empty ();
   bool ok = ResolveTypeToCanonicalPath::go (impl_block.get_type (), self_cpath);
-  rust_assert (ok);
+  if (!ok)
+    {
+      resolver->get_name_scope ().pop ();
+      resolver->get_type_scope ().pop ();
+      resolver->get_label_scope ().pop ();
+      return;
+    }
+
   rust_debug ("AST::InherentImpl resolve Self: {%s}",
 	      self_cpath.get ().c_str ());
 
@@ -671,12 +678,17 @@ ResolveItem::visit (AST::TraitImpl &impl_block)
       return;
     }
 
-  bool ok;
   // setup paths
   CanonicalPath canonical_trait_type = CanonicalPath::create_empty ();
-  ok = ResolveTypeToCanonicalPath::go (impl_block.get_trait_path (),
-				       canonical_trait_type);
-  rust_assert (ok);
+  bool ok = ResolveTypeToCanonicalPath::go (impl_block.get_trait_path (),
+					    canonical_trait_type);
+  if (!ok)
+    {
+      resolver->get_name_scope ().pop ();
+      resolver->get_type_scope ().pop ();
+      resolver->get_label_scope ().pop ();
+      return;
+    }
 
   rust_debug ("AST::TraitImpl resolve trait type: {%s}",
 	      canonical_trait_type.get ().c_str ());
@@ -684,7 +696,13 @@ ResolveItem::visit (AST::TraitImpl &impl_block)
   CanonicalPath canonical_impl_type = CanonicalPath::create_empty ();
   ok = ResolveTypeToCanonicalPath::go (impl_block.get_type (),
 				       canonical_impl_type);
-  rust_assert (ok);
+  if (!ok)
+    {
+      resolver->get_name_scope ().pop ();
+      resolver->get_type_scope ().pop ();
+      resolver->get_label_scope ().pop ();
+      return;
+    }
 
   rust_debug ("AST::TraitImpl resolve self: {%s}",
 	      canonical_impl_type.get ().c_str ());

--- a/gcc/testsuite/rust/compile/issue-2423.rs
+++ b/gcc/testsuite/rust/compile/issue-2423.rs
@@ -1,0 +1,14 @@
+impl NonExistant {
+    // { dg-error "failed to resolve" "" { target *-*-* } .-1 }
+    fn test() {}
+}
+
+impl NotFound for NonExistant {
+    // { dg-error "failed to resolve" "" { target *-*-* } .-1 }
+    fn test() {}
+}
+
+trait A {}
+
+impl A for NotFound {}
+// { dg-error "failed to resolve" "" { target *-*-* } .-1 }

--- a/gcc/testsuite/rust/compile/nr2/exclude
+++ b/gcc/testsuite/rust/compile/nr2/exclude
@@ -206,4 +206,5 @@ issue-1773.rs
 issue-2905-1.rs
 issue-2905-2.rs
 issue-2907.rs
+issue-2423.rs
 # please don't delete the trailing newline

